### PR TITLE
Parser for Effect Handlers

### PR DESCRIPTION
**Pull request by CAG2Mark**
Parse terms of the form
```fs
handle h = Effect with
  fun perform(r) = r(foo)
in bar(h)
```

### CODE REVIEW

**Comment by CAG2Mark** [comment]
Honestly the code right now is pretty unclean and I would prefer to just parse `handle h = Effect with ...` as a `LetLike`, with `handle` being a `letLike` keyword, and `Effect with ...` as an expression which creates an anonymous instance of some trait.

**Comment by LPTK** [comment]
> Honestly the code right now is pretty unclean and I would prefer to just parse `handle h = Effect with ...` as a `LetLike`, with `handle` being a `letLike` keyword, and `Effect with ...` as an expression which creates an anonymous instance of some trait.

It seems fine for now. We can always refactor it later. And it's not like this is necessarily the final syntax either, anyway. We may well change it at some point.

**Comment by CAG2Mark** [comment]
Merged changes from hkmc2

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala [hunk 1]
```diff
@@ -189,6 +191,24 @@ object ParseRule:
...
+          ParseRule("'handle' binding head"):
+            Kw(`=`):
+              ParseRule("'handle' binding equals sign"):
+                Expr(
+                  ParseRule("'handle' binding class name"):
+                    typeDeclTemplateThen(
+                      Kw(`in`):
+                        ParseRule(s"'handle' binding `in` clause")(
+                          exprOrBlk(ParseRule(s"'handle' binding body")(End(())))((body, _: Unit) => S(body))
+                      *),
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: It seems this line is mis-indented.

File: hkmc2/shared/src/test/mlscript/parser/Handler.mls [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  fun f(r) = r(0)
+in
+  foo(h)
+//│ ╔══[ERROR] Unsupported handle binding shape
+//│ ║  l.38: 	handle 1 = 1 with
+//│ ║        	       ^^^^^^^^^^
+//│ ║  l.39: 	  fun f(r) = r(0)
+//│ ╙──      	^^^^^^^^^^^^^^^^^
+
+:elt
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/parser/Handler.mls]: Let's omit this flag from the test file. It generates mostly noise.

File: hkmc2/shared/src/test/mlscript/parser/Handler.mls [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+handle h = Eff
+//│ ╔══[PARSE ERROR] Expected 'with' keyword after 'handle' binding class name; found end of input instead
+//│ ║  l.16: 	handle h = Eff
+//│ ╙──      	              ^
+
+:pe
+handle h = Eff with
+  fun f(r) = r(0)
+  fun g(a, r) = r(1)
+//│ ╔══[PARSE ERROR] Expected 'in' keyword after type declaration block; found end of input instead
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/parser/Handler.mls]: Why not allow a statement form? Just like `let` – where both are valid:
```fs
fun f() =
  let x = 1 in
    log(x)

// or

fun f() =
  let x = 1
  log(x) 
```

**Review by LPTK — APPROVED** [review]
Ok thanks, please address my minor comments, and then I'll make the merge!

File: hkmc2/shared/src/test/mlscript/parser/Handler.mls (hunk 2 above)
Comment (CAG2Mark) [reply hkmc2/shared/src/test/mlscript/parser/Handler.mls]: Is there a way to run the elaborator without generating all the noise? It seems to do basic name analysis and this is something we would should test.

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala (hunk 1 above)
Comment (CAG2Mark) [reply hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: oh it seems just the last comma is misindented, will be fixed when I figure out the other change I'm working on

Comment (LPTK) [reply hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: No, it's the whole thing. This `*)` corresponds to the `(` in `ParseRule(s"'handle' binding `in` clause")(` and should be indented to the same level.

File: hkmc2/shared/src/test/mlscript/parser/Handler.mls (hunk 2 above)
Comment (LPTK) [reply hkmc2/shared/src/test/mlscript/parser/Handler.mls]: The elaborator always runs even if you're not printing its result. Just remove the flag.

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala (hunk 1 above)
Comment (CAG2Mark) [reply hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: Oh then the problem exists in `letLike` as well which I copy-pasted from. I'll fix that as well

**Comment by CAG2Mark** [comment]
All comments addressed, please see the latest changes.

File: hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala (hunk 1 above)
Comment (LPTK) [reply hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala]: Probably. Feel free to fix it there as well! 😉

File: hkmc2/shared/src/test/mlscript/HkScratch.mls [hunk 4]
```diff
@@ -4,4 +4,9 @@
 // :elt
 
 
-
+:parseOnly
+(let x = 1) + 1
+x
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/HkScratch.mls]: This should be removed ~~or moved to another test file. Probably `hkmc2/shared/src/test/mlscript/parser/Let.mls`.~~ EDIT: I see it's already in `codegen/Bindings.mls`.

Comment (LPTK) [reply hkmc2/shared/src/test/mlscript/HkScratch.mls]: (We try to keep the scratch file empty.)

Comment (CAG2Mark) [reply hkmc2/shared/src/test/mlscript/HkScratch.mls]: my bad, forgot to remove it before my commit >_>

File: hkmc2/shared/src/test/mlscript/HkScratch.mls [hunk 5]
```diff
@@ -2,6 +2,3 @@
 :sjs
 // :pt
 // :elt
-
-
-
```
Comment (LPTK) [inline hkmc2/shared/src/test/mlscript/HkScratch.mls]: Please do not commit any unnecessary whitespace changes.

**Comment by LPTK** [comment]
Pro-tip: if you had reviewed your diff first, these repeated rounds of comments on silly things would have been avoided 😉 

https://github.com/hkust-taco/mlscript/blob/mlscript/CONTRIBUTING.md#for-non-draft-prs

**Comment by CAG2Mark** [comment]
Fixed
